File: src/SOP8/Egovframework/Base/base-account-bll/src/main/java/egovframework/base/account/bll/process/login/ExternalLoginManager.java
package egovframework.base.account.bll.process.login;

import egovframework.base.account.ibll.models.ApplicationUser;
import egovframework.base.account.ibll.response.LoginResult;
import egovframework.base.model.account.Grade;
import egovframework.base.model.account.User;
import egovframework.base.model.account.Session;
import egovframework.base.model.common.Site;
import org.json.JSONObject;
import org.json.JSONArray;
import org.json.JSONException;
import egovframework.base.account.bll.resource.ErrorMessage;
import egovframework.base.response.resource.ID;
import egovframework.base.account.bll.process.LoadManager;
import egovframework.base.generic.Holder;

import egovframework.base.dal.account.*;
import egovframework.base.dal.common.SiteDAO;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.*;

public class ExternalLoginManager {

    private GradeDAO gradeDao;
    private UserDAO userDao;
    private SiteDAO siteDao;
    private SessionDAO sessionDao;
    //private IDataManager dataManager;

    public ExternalLoginManager(GradeDAO gradeDao, UserDAO userDao, SiteDAO siteDao, SessionDAO sessionDao) {
        this.gradeDao = gradeDao;
        this.userDao = userDao;
        this.siteDao = siteDao;
        this.sessionDao = sessionDao;
        //this.dataManager = dataManager;
    }

    public LoginResult externalLogin(String userID, String password, String externalLoginUrl,
                                     String sessionKey, boolean autoLogin) {
        try {
            JSONObject jsonData = new JSONObject();
            jsonData.put("userID", userID);
            jsonData.put("hashCode", password);

            JSONObject jsonWrapper = new JSONObject();
            jsonWrapper.put("externalLogin", jsonData);

            String jsonString = jsonWrapper.toString();
            byte[] bytes = jsonString.getBytes(StandardCharsets.UTF_8);

            HttpURLConnection conn = (HttpURLConnection) new URL(externalLoginUrl).openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
            conn.setDoOutput(true);
            conn.getOutputStream().write(bytes);

            BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
            StringBuilder responseBuilder = new StringBuilder();
            String line;

            while ((line = reader.readLine()) != null) {
                responseBuilder.append(line.trim());
            }

            reader.close();
            conn.disconnect();

            return getExternalLoginResult(responseBuilder.toString(), sessionKey, autoLogin);
        } catch (IOException e) {
            return new LoginResult(false, e.getMessage());
        }
    }

    private LoginResult getExternalLoginResult(String response, String sessionKey, boolean autoLogin) {
        try {
            JSONObject json = new JSONObject(response);

            String userID = json.optString("userID");
            String userName = json.optString("name");
            String teamName = json.optString("teamName");
            String _errorMessage = json.optString("message");
            String successStr = json.optString("success", "false").toLowerCase();

            if (userID.isEmpty() || userName.isEmpty() || !successStr.equals("true")) {
                return new LoginResult(false, _errorMessage);
            }

            Holder<String> errorMessage = new Holder<String>();

            // 등급 조회
            List<Grade> grades = this.gradeDao.selectGrades(null, null, null, errorMessage);
            //List<Grade> grades = dataManager.getSelect().select(Grade.class);
            if (grades == null) {
                return new LoginResult(false, errorMessage.value);
            }

            if (grades.isEmpty()) {
                return new LoginResult(false, ID.get(ErrorMessage.class, "noneUserLevel").value());
            }

            Grade grade = grades.get(0);

            // 사용자 정보 조회 또는 생성
            String condition = String.format("%s = '%s'", User.Fields.user_id, userID);
            User user = this.userDao.selectUser(null, condition, errorMessage);
            //User user = dataManager.getSelect().selectFirst(User.class, condition);

            if (user == null) {
                user = new User();
                user.setGrad_sn(grade.getGrad_sn());
                user.setPassword("");
                user.setUser_id(userID);
                user.setUser_name(userName);
                user.setPassword_salt(LoadManager.makeSalt());
                user.setSite_sn(getFirstSiteNo());

                //int addedId = dataManager.getCreate().insert(user);
                if (!this.userDao.insertUser(user, errorMessage)) {
                    return new LoginResult(false, ID.get(ErrorMessage.class, "failToCreateUser").value());
                }

                //user.user_sn = addedId;
            }

            if (!LoginManager.updateSession(this.sessionDao, user.getUser_sn(), sessionKey, autoLogin, errorMessage)) {
                return new LoginResult(false, ID.get(ErrorMessage.class, "failToSession").value());
            }

            ApplicationUser appUser = new ApplicationUser();
            appUser.setUserSn(user.getUser_sn());
            appUser.setGradName(grade.getGrad_name());
            appUser.setGradSn(grade.getGrad_sn());
            appUser.setUserName(userName);
            appUser.setUserId(userID);
            appUser.setSessionKey(sessionKey);
            appUser.setSiteSn(user.getSite_sn());
            appUser.setRegularMemberNo(user.getRgl_memb_sn());

            LoginResult result = new LoginResult(true, "");
            result.setUser(appUser);
            return result;

        } catch (JSONException e) {
            return new LoginResult(false, "JSON parsing error: " + e.getMessage());
        }
    }

    private Integer getFirstSiteNo() {
        Holder<String> errorMessage = new Holder<String>();
        List<Site> sites = this.siteDao.selectSites(null, null, null, errorMessage);
        //List<Site> sites = dataManager.getSelect().select(Site.class);
        if (sites == null || sites.isEmpty()) {
            return null;
        }
        return sites.get(0).getSite_sn();
    }
}

File: src/SOP8/Base/Account/Base.Account.BLL/Process/Login/ExternalLoginManager.cs
using System;
using System.Collections.Generic;
using dnsDapperDBUtil.DataAccessLayer.IDAL;
using Base.Model.Account;
using Base.Account.IBLL.Models;
using Base.Account.IBLL.Response;
using Newtonsoft.Json.Linq;
using System.IO;
using Response.Resource;

namespace Base.Account.BLL.Process.Login
{
    using Resource;

    class ExternalLoginManager
    {
        private IDataManager m_dataManager = null;

        public ExternalLoginManager(IDataManager dataManager)
        {
            m_dataManager = dataManager;
        }

        public LoginResult ExternalLogin(string strUserID, string strPW, string strExternalLoginURL, string strSessionKey, bool autoLogin, out User user, out Grade grade)
        {
            user = null;
            grade = null;

            JObject json = new JObject();
            json.Add("account", strUserID);
            json.Add("password", strPW);

            string strJson = json.ToString();

            byte[] bytes = System.Text.Encoding.UTF8.GetBytes(strJson);
            int len = bytes.Length;

            System.Net.HttpWebRequest request = (System.Net.HttpWebRequest)System.Net.WebRequest.Create(new Uri(strExternalLoginURL));
            request.Method = "POST";
            request.ContentType = "application/json; charset=utf-8";
            request.ContentLength = len + 3;

            string strResult = "";
            string strErrorMessage = null;
            LoginResult result = new LoginResult();

            try
            {
                StreamWriter writer = new StreamWriter(request.GetRequestStream(), System.Text.Encoding.UTF8);
                writer.Write(strJson);
                writer.Close();

                System.Net.HttpWebResponse wRes = (System.Net.HttpWebResponse)request.GetResponse();

                Stream respPostStream = wRes.GetResponseStream();
                StreamReader readerPost = new StreamReader(respPostStream, System.Text.Encoding.UTF8);

                strResult = readerPost.ReadToEnd().Trim();
                request.Abort();
                readerPost.Close();
                respPostStream.Close();
                strErrorMessage = null;

                return GetExternalLoginResult(strResult, strUserID, strSessionKey, autoLogin);
            }
            catch (System.Net.WebException ex)
            {
                strErrorMessage = ex.Message;
            }

            result.Success = false;
            result.Message = strErrorMessage;
            return result;
        }

        private LoginResult GetExternalLoginResult(string strResult, string strUserID, string strSessionKey, bool autoLogin = false)
        {
            string strErrorMessage;
            string strUserName, strTeamName;
            bool success = GetJsonResult(JObject.Parse(strResult), out strUserName, out strTeamName, out strErrorMessage);

            if (success == false)
                return new LoginResult(false, strErrorMessage);

            LoginResult result = new LoginResult();

            IEnumerable<Grade> grades = m_dataManager.GetSelect().Select<Grade>(null, out strErrorMessage);

            if (grades == null)
            {
                result.Success = false;
                result.Message = strErrorMessage;
                return result;
            }

            Grade grade = null;

            foreach (Grade _grade in grades)
            {
                grade = _grade;
                break;
            }

            if (grade == null)
            {
                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("noneUserLevel").Value();
                return result;
            }

            Dictionary<User.Fields, object> dicConditions = new Dictionary<User.Fields, object>();
            dicConditions[User.Fields.user_id] = strUserID;

            string strCondition = string.Format("{0} = '{1}'", User.Fields.user_id, strUserID);
            User user = m_dataManager.GetSelect().SelectFirst<User>(strCondition, out strErrorMessage);

            if (user == null)
            {
                if (strErrorMessage != null)
                {
                    System.Diagnostics.Trace.WriteLine(strErrorMessage);
                    return new LoginResult(false, ID.Get<ErrorMessage>("unknownUser").Value());
                }
                else
                {
                    // 없으면 새로 생성
                    user = new User();

                    user.grad_sn = grade.grad_sn;
                    user.password = "";
                    user.user_id = strUserID;
                    user.user_name = strUserName;
                    user.password_salt = LoadManager.MakeSalt();
                    user.site_sn = GetFirstSiteNo();

                    int addedID;

                    if (m_dataManager.GetCreate().Insert<User>(user, out addedID, out strErrorMessage) == false)
                    {
                        System.Diagnostics.Trace.WriteLine(strErrorMessage);

                        result.Success = false;
                        result.Message = ID.Get<ErrorMessage>("failToCreateUser").Value();
                        return result;
                    }

                    user.user_sn = addedID;
                }
            }

            grade = null;

            foreach (Grade _grade in grades)
            {
                if (user.grad_sn < 0)
                {
                    grade = _grade;
                    break;
                }
                else
                {
                    if (_grade.grad_sn == user.grad_sn)
                    {
                        grade = _grade;
                        break;
                    }
                }
            }

            if (grade == null)
            {
                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("noneUserLevel").Value();
                return result;
            }

            if (LoginManager.UpdateSession(m_dataManager, user.user_sn, strSessionKey, autoLogin, out strErrorMessage) == false)
            {
                System.Diagnostics.Trace.WriteLine(strErrorMessage);

                result.Success = false;
                result.Message = ID.Get<ErrorMessage>("failToSession").Value();
                return result;
            }

            ApplicationUser loginUser = new ApplicationUser();
            loginUser.user_sn = user.user_sn;
            loginUser.grad_name = grade.grad_name;
            loginUser.grad_sn = grade.grad_sn;
            loginUser.user_name = strUserName;
            loginUser.user_id = strUserID;
            loginUser.session_key = strSessionKey;
            loginUser.site_sn = user.site_sn;
            loginUser.RegularMemberNo = user.rgl_memb_sn;

            result.Success = true;
            result.Message = "";
            result.User = loginUser;
            return result;
        }

        private int? GetFirstSiteNo()
        {
            string strErrorMessage;
            IEnumerable<Model.Common.Site> sites = m_dataManager.GetSelect().Select<Model.Common.Site>(null, out strErrorMessage);

            if (sites == null)
            {
                System.Diagnostics.Trace.WriteLine("GetFirstSiteNo Fail : " + strErrorMessage);
                return null;
            }

            foreach (var site in sites)
            {
                return site.site_sn;
            }

            return null;
        }

        private bool GetJsonResult(JObject json, out string strUserName, out string strTeamName, out string strErrorMessage)
        {
            strUserName = strTeamName = null;
            strErrorMessage = null;

            if (json == null)
                return false;

            JToken tokenUser = json.GetValue("user");

            JToken tokenName = ((JObject)tokenUser).GetValue("user_name");
            JToken tokenTeamName = ((JObject)tokenUser).GetValue("user_division");
            JToken tokenSuccess = json.GetValue("success");

            if (tokenName == null || tokenTeamName == null)
                return false;

            strUserName = tokenName.Value<string>();

            if (tokenTeamName != null)
                strTeamName = tokenTeamName.Value<string>();

            if (tokenSuccess != null)
            {
                string strSuccess = tokenSuccess.Value<string>().ToLower();

                if (strSuccess == "true")
                    return true;
            }

            strErrorMessage = "ID 또는 비밀번호를 확인하세요.";
            return false;
        }
    }
}
